torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,602 @@
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, tolerance_grad);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, tolerance_change);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, history_size);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG_OPTIONAL(std::string, line_search_fn);
+}
+
+template <typename T>
+bool if_container_equal(T lhs, T rhs) {
+  if (!(lhs.size() == rhs.size())) return false;
+  for (int i = 0; i < lhs.size(); i++) {
+    if (!torch::equal(lhs.at(i), rhs.at(i))) return false;
+  }
+  return true;
+}
+
+bool operator==(const LBFGSParamState& lhs, const LBFGSParamState& rhs) {
+  return (lhs.func_evals() == rhs.func_evals()) &&
+         (lhs.n_iter() == rhs.n_iter()) &&
+         (lhs.t() == rhs.t()) &&
+         torch::equal_if_defined(lhs.d(), lhs.d()) &&
+         torch::equal_if_defined(lhs.H_diag(), rhs.H_diag()) &&
+         torch::equal_if_defined(lhs.prev_flat_grad(), rhs.prev_flat_grad()) &&
+         torch::equal_if_defined(lhs.prev_loss(), rhs.prev_loss()) &&
+         if_container_equal(lhs.old_dirs(), rhs.old_dirs()) &&
+         if_container_equal(lhs.old_stps(), rhs.old_stps()) &&
+         if_container_equal(lhs.ro(), rhs.ro()) &&
+         if_container_equal(lhs.al(), rhs.al());
+}
+
+void LBFGSParamState::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(func_evals);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(n_iter);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(t);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(d);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(H_diag);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(prev_flat_grad);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(prev_loss);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG_DEQUE(old_dirs);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG_DEQUE(old_stps);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG_DEQUE(ro);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(al);
+}
+
+void LBFGSParamState::serialize(torch::serialize::InputArchive& archive) {
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, func_evals);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, n_iter);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, t);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, d);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, H_diag);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, prev_flat_grad);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, prev_loss);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG_DEQUE(std::deque<Tensor>, old_dirs);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG_DEQUE(std::deque<Tensor>, old_stps);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG_DEQUE(std::deque<Tensor>, ro);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(std::vector<Tensor>, al);
+}
+
+Tensor LBFGS::_gather_flat_grad() {
   std::vector<Tensor> views;
-  for (auto& parameter : parameters_) {
-    if (!parameter.grad().defined()) {
-      views.push_back(parameter.new_empty({parameter.numel()}).zero_());
+  for (const auto& p : param_groups_.at(0).params()) {
+    Tensor view;
+    if (!p.grad().defined()) {
+      views.emplace_back(p.new_empty({p.numel()}).zero_());
```

Comment:
I think we might want to assign the value to `view`, and then do the emplace_back at the end :D

Reply:
or we can just straight away emplace_back in each branch to avoid an extra copy. I think obviously maintaining as much semantic parity between Python and C++ code is important but we should ensure, we are not compromising optimality because of that

Reply:
removed the line ```Tensor view``` though :D